taichi/codegen/spirv/spirv_codegen.cpp

```diff
@@ -1676,34 +1791,60 @@ class TaskCodegen : public IRVisitor {
   }
 
   void compile_args_struct() {
+    if (!ctx_attribs_->has_args())
+      return;
+
     std::vector<std::tuple<spirv::SType, std::string, size_t>>
         struct_components_;
     for (auto &arg : ctx_attribs_->args()) {
-      struct_components_.emplace_back(ir_->get_primitive_type(arg.dt),
-                                      "arg" + std::to_string(arg.index),
-                                      arg.offset_in_mem);
+      if (arg.is_array &&
+          device_->get_cap(
+              DeviceCapability::spirv_has_physical_storage_buffer)) {
+        struct_components_.emplace_back(ir_->u64_type(),
+                                        "arg_ptr" + std::to_string(arg.index),
+                                        arg.offset_in_mem);
+      } else {
+        struct_components_.emplace_back(ir_->get_primitive_type(arg.dt),
+                                        "arg" + std::to_string(arg.index),
+                                        arg.offset_in_mem);
+      }
+    }
+    // A compromise for use in constants buffer
+    // where scalar arrays follow very weird packing rules
+    for (int i = 0; i < ctx_attribs_->extra_args_bytes() / 4; i++) {
+      struct_components_.emplace_back(
+          ir_->i32_type(), "extra_args" + std::to_string(i),
+          ctx_attribs_->extra_args_mem_offset() + i * 4);
     }
-    size_t args_size = ctx_attribs_->args_bytes();
+    args_struct_type_ = ir_->create_struct_type(struct_components_);
+
+    args_buffer_value_ =
+        ir_->uniform_struct_argument(args_struct_type_, 0, 0, "args");
+  }
+
+  void compile_ret_struct() {
+    if (!ctx_attribs_->has_rets())
+      return;
+
+    std::vector<std::tuple<spirv::SType, std::string, size_t>>
+        struct_components_;
     for (auto &ret : ctx_attribs_->rets()) {
       struct_components_.emplace_back(ir_->get_primitive_type(ret.dt),
                                       "ret" + std::to_string(ret.index),
                                       ret.offset_in_mem);
     }
-    struct_components_.emplace_back(
-        ir_->get_array_type(ir_->i32_type(),
-                            ctx_attribs_->extra_args_bytes() / 4),
-        "extra_args", ctx_attribs_->extra_args_mem_offset());
-    args_struct_type_ = ir_->create_struct_type(struct_components_);
-    args_buffer_value_ =
-        ir_->buffer_struct_argument(args_struct_type_, 0, 0, "args");
+    ret_struct_type_ = ir_->create_struct_type(struct_components_);
+
+    ret_buffer_value_ =
+        ir_->buffer_struct_argument(ret_struct_type_, 0, 1, "rets");
   }
 
   std::vector<BufferBind> get_buffer_binds() {
     std::vector<BufferBind> result;
     for (auto &[key, val] : buffer_binding_map_) {
       result.push_back(BufferBind{key.first, int(val)});
     }
-    return result;
+    return std::move(result);
```

Comment:
We don't need `move`, RVO is guaranteed :-) https://stackoverflow.com/a/17473874/12003165